### PR TITLE
feat: Enhance Ignore files and folders feature

### PR DESCRIPTION
**Pull request by enguerran**
I extracted some code from the class as they can be standalone
functions: easier to test, better insulation.

I reorganized tests by creating categories: easier to maintain,
faster to read.

I extracted default patterns to a standalone configuration file:
easier to patch, better SRP.

Let me know if every thing looks great to maintainers.

### CODE REVIEW

File: core/ignore.js [hunk 1]
```diff
@@ -1,164 +1,110 @@
...
-      'ehthumbs.db'
-    ]
-
-    // See https://github.com/jonschlinkert/micromatch#options
-    MicromatchOptions =
-            {noextglob: true}
-  }
-
-  /*::
-  patterns: IgnorePattern[]
```
Comment (aenario) [inline core/ignore.js]: These comments `/*::` are flow annotation and the reason why this PR is falling CI (https://travis-ci.org/cozy-labs/cozy-desktop/jobs/446757368#L712)

Run `yarn lint:flow` locally to see the errors.

Comment (enguerran) [reply core/ignore.js]: oops, I'll check that.

**Comment by enguerran** [comment]
@aenario I added a commit to handle flow syntax. Could you update your review?

File: core/ignore.js [hunk 2]
```diff
@@ -13,152 +13,119 @@ export type IgnorePattern = {
...
-      if (pattern.match(path)) { return true }
-    }
-    let parent = dirname(path)
-    if (parent === '.') { return false }
-    return this.match(parent, true, pattern)
-  }
-
   // Return true if the given file/folder path should be ignored
-  isIgnored (doc) {
+  isIgnored (doc /*: CozyDoc */) {
```
Comment (sebn) [inline core/ignore.js]: Maybe use the `Metadata` type instead? (as defined in `core/metadata.js`)

Comment (enguerran) [reply core/ignore.js]: I looked for a type annotation that might be suitable for a `cozy-doc' but I couldn't find it. Thanks for letting me know. The naming doesn't seem to me very explicit.

**Comment by enguerran** [comment]
I took advantage of the [type](https://github.com/cozy-labs/cozy-desktop/pull/1229/commits/dfd0f441da9c9c15280955a61621d72ffb967c68#diff-a4bdab4052b01b57644232f7c87f3a33R111) modification [for which you warned me](https://github.com/cozy-labs/cozy-desktop/pull/1229#discussion_r228993998) to rebase all the commits on master.

**Pull request merged**